File: downloads/python/labs2/ISU/Jewels/BasicIcon.py
#!/usr/bin/env python
# coding: UTF-8
#
## @package BasicIcon
#
#   Basic implementation of the Icon interface.
#
#   @author Paulo Roma
#   @since 30/03/2020
#

from Icon import Icon

##
# Basic implementation of the Icon interface.
#
class BasicIcon (Icon):
    ##
    # Constructs an icon of the given type.
    # @param type
    #   type for this icon
    #
    def __init__(self, type):
        ## Type of this icon.
        self.__type = type

    ## Returns the type of this icon.
    # @return type of this icon.
    #
    def getType(self):
        return self.__type

    ## Return whether two icons have the same type.
    def __eq__(self, obj):
        if obj is None or self.__type is None or type(obj) != type(self):
            return False
        return self.__type == obj.__type

    ## Return a string representaion of this icon.
    def __repr__(self):
        return str(self.__type)

File: 3-javascript/jewels/jewels/private/BasicIcon.js
/**  @file
 *
 *   Summary.
 *   <p>Basic implementation of the Icon interface.</p>
 *
 *   @author Paulo Roma
 *   @see <a href="/cwdc/3-javascript/jewels/BasicIcon.js">source</a>
 *   @since 10/03/2021
 */

import { Icon } from "./Icon.js";

/** Basic implementation of the Icon interface.
 *
 */
export class BasicIcon extends Icon {
  #type;

  /** Constructs an icon of the given type.
   *
   *  @param {Number} type
   *      type for this icon
   */
  constructor(type) {
    super();

    /**
     * Type of this icon.
     * @type {Number}
     */
    this.#type = type;
  }

  /** Returns the type of this icon.
   *
   *  @return {Number} type of this icon.
   */
  getType() {
    return this.#type;
  }

  /** Return whether two icons have the same type.
   *
   *  @param {Icon} obj object to compare to.
   *  @return {Boolean} true it the types are the same.
   */
  equal(obj) {
    if (obj == null || this.#type == null || typeof this != typeof obj) {
      return false;
    }

    return this.#type == obj.#type;
  }

  /** Return a string representaion of this icon.
   *
   *  @return {String} this icon representation.
   */
  repr() {
    return self.#type.toString();
  }
}
